torch/export/_decomp_utils.py

```diff
@@ -0,0 +1,178 @@
+    3. If it is write operation, we don't necessarily materialize
+    4. We load the final time during export, right before calling run_decompositions()
+    """
+
+    def __init__(self):
+        super().__init__()
+        from torch._decomp import _core_aten_decompositions_post_autograd
+
+        # For aten ops, we load them up in the beginning
+        self.aten_decomp_table = _core_aten_decompositions_post_autograd()
+
+        for op in _collect_all_valid_cia_ops_for_aten_namespace():
+            self.aten_decomp_table[op] = _get_decomp_for_cia(op)
+
+        # This is to track the *pending* deleted custom ops that haven't been materialized yet
+        self.deleted_custom_ops = set()
+        # When we can materialize a custom op, we use this table
+        self.additional_custom_op_decomp = {}
+        # When this is true, there shouldn't be any pending operations in the table.
+        self.has_materialized = False
+
+    def __getitem__(self, key):
+        self._materialize_if_needed()
+
+        if _is_aten_op(key):
+            return self.aten_decomp_table[key]
+
+        assert _is_custom_op(key)
+        return self.additional_custom_op_decomp[key]
+
+    def __setitem__(self, key, value):
+        if _is_aten_op(key):
+            self.aten_decomp_table[key] = value
+            return
+
+        assert _is_custom_op(key)
+
+        self.additional_custom_op_decomp[key] = value
+
+        if key in self.deleted_custom_ops:
+            self.deleted_custom_ops.remove(key)
+
+    def keys(self):
+        self._materialize_if_needed()
+
+        return itertools.chain(
+            self.aten_decomp_table.keys(), self.additional_custom_op_decomp.keys()
+        )
+
+    def __delitem__(self, key):
+        self.pop(key)
+
+    def update(self, other_dict):  # type: ignore[override]
+        for k, v in other_dict.items():
+            if _is_aten_op(k):
+                self.aten_decomp_table[k] = v
+            else:
+                self.additional_custom_op_decomp[k] = v
+                if not self.has_materialized:
+                    if k in self.deleted_custom_ops:
+                        self.deleted_custom_ops.remove(k)
+
+    def __missing__(self, key):
+        return not self.__contains__(key)
+
+    def __contains__(self, key):
+        self._materialize_if_needed()
+        if _is_aten_op(key):
+            return key in self.aten_decomp_table
+
+        if key in self.additional_custom_op_decomp:
+            return True
+
+        return False
+
+    def __len__(self):
+        self._materialize_if_needed()
+        return len(self.aten_decomp_table) + len(self.additional_custom_op_decomp)
+
+    def __iter__(self):
+        self._materialize_if_needed()
+        return itertools.chain(
+            self.aten_decomp_table.__iter__(),
+            self.additional_custom_op_decomp.__iter__(),
+        )
+
+    def __reverse__(self):
+        raise RuntimeError("Cannot call reverse() on custom decomp table")
```

Comment:
just curious, but why does the default reverse not just work?

Reply:
Ahh this is because we have two tables..